### DISCUSSION TITLE
Integrating steppers into a FormCollection

### DISCUSSION THREAD

**Discussion by mzeba** [jrief/django-formset#86] [Q&A]
Hi, is there a way to make stepper with django-formset? I've tried django-formtools but integrating it into django-formset seems complicated.

**Comment by jrief** [comment]
> Hi, is there a way to make stepper with django-formset?

what do you mean by "stepper"?

**Reply by mzeba** [reply]
I'm thinking of a multi-stage filling process. With 'next' and 'previous' buttons. At the last stage, the form is submitted.

**Reply by mzeba** [reply]
An example: [Bootstrap stepper](https://johann-s.github.io/bs-stepper/)

**Reply by jrief** [reply]
Oh yes, this is a feature I intend to add in the near future because it makes a lot of sense in the context of form collections. Didn't know it also is named "stepper".

If you want helping me to implement this, you could start looking for appropriate styles for TailwindCSS. Having them, you can then create an example on Codepen. In django-angular I have implemented such a feature many years ago.

This however may take some time, because I have other features awaiting to be implemented.

**Reply by mzeba** [reply]
I'm in. I'll do some research.

**Comment by kashyl** [comment]
Hi! I was wondering if anyone had any hints on how one may implement this using django-formset and django-formtools' wizard? Both libraries are great, and I've had success setting nice forms with this one, as well as multi-step forms with formtools, though combining them seems to be quite complicated, as @parice02  said

**Comment by jrief** [comment]
@kashyl
multi-step forms are on my TODO-list and I presumably will add them to version 1.7.

**Comment by jrief** [comment]
@parice02 @kashyl In the current HEAD of the `develop` branch there is a working Form Stepper. I haven't added and documentation yet and currently there is only a demo. Try http://localhost:8000/bootstrap/checkout in your local setup. This feature will be added for sure in version 1.7.

**Reply by jrief** [reply]
This feature has been officially released in version 1.7. Please [check the docs for details](https://django-formset.fly.dev/form-stepper/).

**Comment by Alex-Sichkar** [comment]
Thank you! This is a very useful feature.

I've been testing it on the documentation page and found an issue. After the form is submitted, the user is allowed to navigate back to previous steps, modify the data to an invalid state, then proceed to the last step and submit the form again. However, no validation is triggered for the previous steps' forms.

Is this the intended behavior?

https://github.com/user-attachments/assets/f678e1f2-7470-4461-86f1-a4972ced7828

**Reply by jrief** [reply]
This fixed now and will be available in version 1.7.1.

**Comment by jrief** [comment]
no, this is not intended behaviour. Thanks for reporting, will fix it in the next version.
